pkg/ingester/ingester.go

```diff
@@ -739,6 +832,47 @@ func (i *Ingester) MetricsForLabelMatchers(ctx context.Context, req *client.Metr
 	return result, nil
 }
 
+// MetricMetadata returns all the metric metadata of a user.
+func (i *Ingester) MetricMetadata(ctx context.Context, req *client.MetricMetadataRequest) (*client.MetricMetadataResponse, error) {
+	i.userStatesMtx.RLock()
+	if err := i.checkRunningOrStopping(); err != nil {
+		i.userStatesMtx.RUnlock()
+		return nil, err
+	}
+	i.userStatesMtx.RUnlock()
+
+	userID, err := user.ExtractOrgID(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("no user id")
+	}
+
+	i.usersMetadataMtx.RLock()
+	defer i.usersMetadataMtx.RUnlock()
+	userMetadata, ok := i.usersMetadata[userID]
+	if !ok {
+		return &client.MetricMetadataResponse{}, nil
+	}
+
+	if userMetadata.length() > i.limiter.MaxMetricsWithMetadataPerQuery(userID) {
+		return nil, httpgrpc.Errorf(http.StatusRequestEntityTooLarge, "exceeded maximum number of metadata metrics in a query")
+	}
+
+	metadata := []*client.MetricMetadata{}
+	for _, set := range userMetadata.all() {
+		// If this is a "high-cardinality" metadata set. Let's skip it to at least return
+		// part of the metadata.
+		// TODO: Should we create a metric to keep track of metadata skipped?
+		if set.length() > i.limiter.MaxMetadataPerQuery(userID) {
+			continue
+		}
+		for mm := range set.all() {
+			metadata = append(metadata, mm.toClientMetadata())
+		}
+	}
```

Comment:
If we remove the limit on the read path, we can move all this logic to `metadataMap.toClientMetadata()`. This would have the advantage to remove the `all()` functions which are problematic if we move away from the global lock.